File: backend/src/main/java/dev/adriankluge/hobbyhub/mtg/combo/client/CommanderSpellbookClient.java
package dev.adriankluge.hobbyhub.mtg.combo.client;

import dev.adriankluge.hobbyhub.mtg.combo.dto.ComboResponse;
import java.util.List;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

// Commander Spellbook (backend.commanderspellbook.com) is the actual combo
// database behind EDHREC's own combo pages - querying it directly is more
// reliable than scraping EDHREC's HTML (section 4.4 of the brief). Open, no
// key required. Cached per card name (same Caffeine pattern as Scryfall,
// see CacheConfig) rather than a scheduled Postgres refresh like the news/
// meta features - there's no fixed small set of "top N" results to
// precompute here, every card is its own query.
@Component
public class CommanderSpellbookClient {

    private static final int MAX_COMBOS = 3;
    private static final int MAX_PRODUCES_SHOWN = 3;

    private final RestClient restClient;

    public CommanderSpellbookClient(
            RestClient.Builder restClientBuilder, @Value("${app.mtg.commander-spellbook-base-url}") String baseUrl) {
        this.restClient = restClientBuilder.baseUrl(baseUrl).defaultHeader("Accept", "application/json").build();
    }

    @Cacheable(cacheNames = "commanderSpellbookCombos", key = "#cardName")
    public List<ComboResponse> findCombos(String cardName) {
        String query = "card:\"" + cardName + "\"";
        VariantsResponseDto response = restClient
                .get()
                .uri(uriBuilder -> uriBuilder
                        .path("/variants/")
                        .queryParam("q", query)
                        .queryParam("limit", MAX_COMBOS)
                        .build())
                .retrieve()
                .body(VariantsResponseDto.class);

        if (response == null || response.results() == null) {
            return List.of();
        }
        return response.results().stream().map(variant -> toComboResponse(variant, cardName)).toList();
    }

    private ComboResponse toComboResponse(VariantDto variant, String searchedCardName) {
        List<UsesDto> uses = variant.uses() == null ? List.of() : variant.uses();
        List<String> otherCards = uses.stream()
                .map(use -> use.card().name())
                .filter(name -> !name.equalsIgnoreCase(searchedCardName))
                .toList();
        List<String> produces = variant.produces() == null
                ? List.of()
                : variant.produces().stream().map(p -> p.feature().name()).limit(MAX_PRODUCES_SHOWN).toList();

        return new ComboResponse(
                otherCards, uses.size(), variant.popularity(), produces, "https://commanderspellbook.com/combo/"
                        + variant.id() + "/");
    }
}

File: api/lib/CommanderSpellbookClient.php
<?php
require_once __DIR__ . '/db.php';
require_once __DIR__ . '/http_client.php';
require_once __DIR__ . '/Cache.php';

// Commander Spellbook (backend.commanderspellbook.com) - the actual combo
// database behind EDHREC's own combo pages, open, no key required. Cached
// per card name (cache-aside, 1h TTL - combos change far less often than
// Scryfall search results do) rather than a scheduled batch job, since
// every card is its own query with no fixed small set to precompute.
class CommanderSpellbookClient
{
    private const MAX_COMBOS = 3;
    private const MAX_PRODUCES_SHOWN = 3;
    private const CACHE_TTL_SECONDS = 3600;

    /** @var callable */
    private $fetch;

    // The User-Agent matters: Commander Spellbook sits behind a CDN that
    // answers 403 to header-less requests from some hosts, which is what a
    // failed lookup looks like from here (null, indistinguishable from
    // "no combos").
    public function __construct(?callable $fetch = null)
    {
        $this->fetch = $fetch ?? fn(string $url) => http_get_json($url, 10, ['User-Agent: ' . SCRYFALL_USER_AGENT]);
    }

    // Null means the lookup failed - distinct from [] ("this card has no
    // combos"), which is a real answer worth caching.
    public function findCombos(string $cardName): ?array
    {
        return cache_aside('commander_spellbook_cache', 'card_name', $cardName, self::CACHE_TTL_SECONDS, function () use ($cardName) {
            $query = 'card:"' . $cardName . '"';
            $json = ($this->fetch)(COMMANDER_SPELLBOOK_BASE_URL . '/variants/?' . http_build_query([
                'q' => $query,
                'limit' => self::MAX_COMBOS,
            ]));

            // Without this, an upstream failure collapses into an empty
            // combo list and cache_aside() stores it as if it were data -
            // pinning "no combos" for the full hour, over and over.
            if ($json === null) {
                return null;
            }

            $combos = [];
            foreach (($json['results'] ?? []) as $variant) {
                $combos[] = $this->toCombo($variant, $cardName);
            }
            return $combos;
        });
    }

    private function toCombo(array $variant, string $searchedCardName): array
    {
        $uses = $variant['uses'] ?? [];
        $otherCards = [];
        foreach ($uses as $use) {
            $name = $use['card']['name'] ?? '';
            if (strcasecmp($name, $searchedCardName) !== 0) {
                $otherCards[] = $name;
            }
        }

        $produces = array_slice(
            array_map(fn($p) => $p['feature']['name'] ?? '', $variant['produces'] ?? []),
            0,
            self::MAX_PRODUCES_SHOWN
        );

        return [
            'otherCards' => $otherCards,
            'cardCount' => count($uses),
            'numDecks' => $variant['popularity'] ?? null,
            'produces' => $produces,
            'url' => 'https://commanderspellbook.com/combo/' . ($variant['id'] ?? '') . '/',
        ];
    }
}
